File: src/layout.rs
use std::sync::Arc;
use std::iter::FromIterator;
use std::ops::{AddAssign, Mul};
use std::fmt::Debug;
use output::{Output};

// to flex or not to flex?
#[allow(unused_variables)]
pub trait Flex : Debug + Sync + Send {
    fn stretch(&self, line_width: f32) -> f32 { 0.0 }
    fn shrink(&self, line_width: f32) -> f32 { 0.0 }
    fn width(&self, line_width: f32) -> f32;
    fn height(&self, line_width: f32) -> f32 { 0.0 }
    
    fn measure(&self, line_width: f32) -> FlexMeasure {
        FlexMeasure {
            shrink:     self.shrink(line_width),
            stretch:    self.stretch(line_width),
            width:      self.width(line_width),
            height:     self.height(line_width)
        }
    }
    
    fn flex(&self, factor: f32) -> FlexMeasure {
        let w = self.width(0.);
        FlexMeasure {
            width: w,
            shrink: w / factor,
            stretch: w * factor,
            height: self.height(0.)
        }
    }
}

#[derive(Copy, Clone, Debug)]
pub struct FlexMeasure {
    shrink:     f32,
    stretch:    f32,
    width:      f32,
    height:     f32
}

impl FlexMeasure {
    pub fn zero() -> FlexMeasure {
        FlexMeasure {
            width: 0.,
            stretch: 0.,
            shrink: 0.,
            height: 0.
        }        
    }
    /// factor = -1 => self.shrink,
    /// factor =  0 => self.width,
    /// factor = +1 => self.stretch
    pub fn at(&self, factor: f32) -> f32 {
        (if factor < 0. {
            (self.width - self.shrink)
        } else {
            (self.stretch - self.width)
        } * factor + self.width)
    }
}
impl AddAssign for FlexMeasure {
    fn add_assign(&mut self, rhs: FlexMeasure) {
        self.width += rhs.width;
        self.stretch += rhs.stretch;
        self.shrink += rhs.shrink;
        self.height = self.height.max(rhs.height);
    }
}
impl Mul<f32> for FlexMeasure {
    type Output = FlexMeasure;
    
    fn mul(&self, f: f32) -> FlexMeasure {
        FlexMeasure {
            width:      self.width * f,
            stretch:    self.stretch * f,
            shrink:     self.shrink * f,
            height:     self.height
        }        
    }
}

/// Send is reqired
/// All indexing has to be relative
/// (to enable inserting, replacing and deleting parts of the steam.)
///
/// As simple "space" (_) is the most common item in the stream,
/// is should be represented by only one item.
#[derive(Clone, Debug)]
enum StreamItem<O: Output> {
    /// A single word (sequence of glyphs)
    Word(O::Word),
    
    /// Continue on the next line (fill)
    Linebreak(bool),
    
    /// (breaking, measure)
    Space(bool, FlexMeasure),
    
    /// Somtimes there are different possiblites of representing something.
    /// A Branch solves this by splitting the stream in two parts.
    /// The default path is taken by skipping the specified amount of entries.
    /// The other one by following the next items.
    ///
    /// normal items
    /// BranchEntry(3)
    ///   branched item 1
    ///   branched item 2
    /// BranchExit(1)
    ///   normal item 1
    /// both sides joined here
    BranchEntry(usize),
    
    /// Each BranchEntry is followed by BranchExit. It specifies the number of
    /// items to skip.
    BranchExit(usize)
}


#[derive(Clone, Debug)]
pub struct TokenStream<O: Output> {
    buf:    Vec<StreamItem<O>>,
    space:  bool
}
impl<O: Output> TokenStream<O> {
    pub fn new() -> TokenStream<O> {
        TokenStream {
            buf:    vec![],
            space:  false
        }
    }
    
    pub fn len(&self) -> usize {
        self.buf.len()
    }
    
    fn add(&mut self, i: StreamItem<O>) {
        self.buf.push(i);
    }
    
    pub fn extend(&mut self, t: &TokenStream<O>) {
        self.buf.extend_from_slice(&t.buf);
    }
    
    pub fn word(&mut self, w: O::Word) {
        self.add(StreamItem::Word(w));
    }
    
    pub fn space(&mut self, breaking: bool, measure: FlexMeasure) {
        self.add(StreamItem::Space(breaking, measure));
    }
    
    pub fn newline(&mut self, fill: bool) {
        self.add(StreamItem::Linebreak(fill));
    }
    
    pub fn branch(&mut self, a: &TokenStream<O>, b: &TokenStream<O>) {
        self.add(StreamItem::BranchEntry(b.buf.len() + 1));
        self.extend(b);
        self.add(StreamItem::BranchExit(a.buf.len()));
        self.extend(a);
    }
    
    pub fn branch_many<I>(&mut self, default: TokenStream<O>, others: I)
    where I: Iterator<Item=TokenStream<O>> {
        let mut others: Vec<TokenStream<O>> = others.collect();
        
        if others.len() == 0 {
            self.extend(&default);
            return;
        }
        
        while others.len() > 1 {
            for n in 0 .. others.len() / 2 {
                let mut merged = TokenStream::new();
                {
                    let b = others.pop().unwrap();
                    let ref a = others[n];
                    merged.branch(a, &b);
                }
                others[n] = merged;
            }
        }
        self.branch(&default, &others[0]);
    }
}

#[derive(Copy, Clone, Debug, Default)]
struct LineBreak {
    prev:   usize,
    path:   u64, // one bit for each branch taken (1) or not (0)
    factor: f32,
    score:  f32
}
type Entry = Option<LineBreak>;

pub struct ParagraphLayout<'a, O: Output> {
    items:      &'a Vec<StreamItem<O>>,
    width:      f32,
}
pub struct Line<O: Output> {
    pub words:  Vec<(Arc<O::Word>, f32)>,
    pub height: f32
}

struct LineContext {
    measure:    FlexMeasure,
    path:       u64, // one bit for each branch on this line
    begin:      usize, // begin of line or branch
    pos:        usize, // calculation starts here
    score:      f32, // score at pos
    branches:   u8 // number of branches so far (<= 64)
}

impl<'a, O> ParagraphLayout<'a, O> where O: Output {
    pub fn new(s: &'a TokenStream<O>, width: f32) -> ParagraphLayout<'a, O> {
        ParagraphLayout {
            items: &s.buf,
            width: width
        }
    }
    
    pub fn run(&mut self) -> Vec<Line<O>> {
        use std::iter::repeat;
        
        let limit = self.items.len();
        let mut nodes: Vec<Entry> = repeat(None).take(limit+1).collect();
        nodes[0] = Some(LineBreak::default());
        let mut last = 0;
        
        for start in 0 .. limit {  
            let node = nodes[start];
            match node {
                Some(b) => {
                    last = self.complete_line(
                        &mut nodes,
                        LineContext {
                            measure:    FlexMeasure::zero(),
                            path:       0,
                            score:      b.score,
                            begin:      start,
                            pos:        start,
                            branches:   0
                        }
                    );
                },
                None => {}
            }
        }
        /*
        for (n, node) in nodes.iter().take(limit).enumerate() {
            println!("{:4}  {:?}", n, node);
            println!("      {:?}", self.items[n]);
        }
        println!("{:4}  {:?}", limit, nodes[limit]);
        println!("last: {}", last);
        */
        if last == 0 {
            return vec![];
        }
        
        let mut steps = vec![];
        
        while last > 0 {
            //println!("node {:3} {:?}", end, self.nodes[end]);
            match nodes[last] {
                Some(b) => {
                    //println!("Line from {} to {}", start, end-1);
                    steps.push((b, last-1));
                    last = b.prev;
                },
                _ => unreachable!()
            }
        }
        
        let mut lines = Vec::with_capacity(steps.len());
        for &(b, end) in steps.iter().rev() {
            let mut measure = FlexMeasure::zero();
            let mut words = vec![];
            let mut pos = b.prev;
            let mut branches = 0;
            while pos < end {
                let node = self.items[pos].clone();
                match node {
                    StreamItem::Word(w) => {
                        let x = measure.at(b.factor);
                        measure += w.measure(self.width);
                        words.push((w, x));
                    },
                    StreamItem::Space(_, s) => {
                        measure += s.measure(self.width)
                    },
                    StreamItem::BranchEntry(len) => {
                        if b.path & (1<<branches) == 0 {
                            // not taken
                            pos += len;
                        }
                        branches += 1;
                    },
                    StreamItem::BranchExit(skip) => pos += skip,
                    StreamItem::Newline(_) => unreachable!()
                }
                pos += 1;
            }
            
            lines.push(Line {
                height: measure.height,
                words:  words
            });
        }
        
        lines
    }
    
    fn complete_line(&self, nodes: &mut Vec<Entry>, c: LineContext) -> usize {
        let mut last = c.begin;
        let mut c = c;
        
        while c.pos < self.items.len() {
            let n = c.pos;
            let ref item = self.items[n];
            match item {
                &StreamItem::Word(ref w) => {
                    c.measure += w.measure(self.width);
                },
                &StreamItem::Space(breaking, ref s) => {
                    if breaking {
                        // breaking case:
                        // width is not added yet!
                        if self.maybe_update(&c, &mut nodes[n+1]) {
                            last = n+1;
                        }
                    }
                    
                    // add width now.
                    c.measure += s.measure(self.width);
                }
                &StreamItem::Linebreak(fill) => {
                    use std::cmp::max;
                    
                    if fill {
                        c.measure.width = max(self.width, c.measure.width);
                        c.measure.stretch = max(self.width, c.measure.stretch);
                    }
                
                    if self.maybe_update(&c, &mut nodes[n+1]) {
                        last = n+1;
                    }
                    break;
                },
                &StreamItem::BranchEntry(len) => {
                    use std::cmp;
                    // b 
                    let b_last = self.complete_line(
                        nodes,
                        LineContext {
                            pos:        n + 1,
                            path:       c.path | (1 << c.branches),
                            branches:   c.branches + 1,
                            ..          c
                        }
                    );
                    last = cmp::max(last, b_last);
                    
                    // a follows here
                    c.pos += len;
                    c.branches += 1;
                },
                &StreamItem::BranchExit(skip) => {
                    c.pos += skip;
                }
            }
            
            if c.measure.shrink > self.width {
                break; // too full
            }
            
            c.pos += 1;
        }
        
        last
    }
    
    
    fn maybe_update(&self, c: &LineContext, node: &mut Entry) -> bool {
        let width = self.width;
        let ref m = c.measure;
        
        if width < m.shrink || m.stretch <= m.width {
            return false;
        }
    
        let delta = width - m.width; // d > 0 => stretch, d < 0 => shrink
        let factor = delta / (if delta >= 0. { m.stretch - m.width } else { m.width - m.shrink });
        
        let break_score = c.score - factor * factor;
        match *node {
            None => {
                *node = Some(LineBreak {
                    prev:   c.begin,
                    path:   c.path,
                    factor: factor,
                    score:  break_score
                } );
            },
            Some(other)
            if break_score > other.score => {
                *node = Some(LineBreak {
                    prev:   c.begin,
                    path:   c.path,
                    factor: factor,
                    score:  break_score
                } );
            },
            _ => {}
        }
        true
    }
}

File: data/layout.js
let layout_items;

function layout(items) {
    layout_items = items;
}

let config = {
    space_shrink:   0.5,
    space_width:    1.0,
    space_stretch:  2.0,
    text_width:     30
};
document.addEventListener("DOMContentLoaded", function() {
    let p = document.getElementById("controls");
    
    add_control(p, "space_shrink", 0.2, 1.0, 0.05);
    add_control(p, "space_width", 0.6, 2.0, 0.2);
    add_control(p, "space_stretch", 1.0, 5.0, 0.25);
    add_control(p, "text_width", 10, 60, 1, function(v) {
        document.getElementById("target").style.width = v + "em";
    });
    
    update_layout();
}, true);

function add_control(p, name, min, max, step, callback) {
    let i = document.createElement("input");
    i.setAttribute("type", "range");
    i.setAttribute("min",  min);
    i.setAttribute("max", max);
    i.setAttribute("step", step);
    
    let value = localStorage[name];
    if (value != undefined) {
        i.setAttribute("value", value);
        config[name] = value;
        if (callback != undefined) callback(value);
    }
    
    i.addEventListener("input", function(e) {
        let v = e.target.value;
        config[name] = v;
        localStorage[name] = v;
        console.log(name, "=", v);
        if (callback != undefined) callback(v);
        update_layout();
    }, false);
    p.appendChild(i);
}
function update_layout() {
    let items = layout_items;
    
    let parent = document.getElementById("target");
    while (parent.firstChild) {
        parent.removeChild(parent.firstChild);
    }
    
    let space = document.createElement("span");
    space.innerHTML = "&nbsp;";
    parent.appendChild(space);
    let space_width = space.getBoundingClientRect().width;
    parent.removeChild(space);
    
    let context = {
        word:   function(word)
                {
                    let span = document.createElement("span");
                    span.appendChild(document.createTextNode(word));
                    parent.appendChild(span);
                    let rect = span.getBoundingClientRect();
                    parent.removeChild(span);
                    return {
                        shrink:     rect.width,
                        width:      rect.width,
                        stretch:    rect.width,
                        height:     rect.height
                    };
                },
        space:  function(scale)
                {
                    let width = space_width * scale;
                    return {
                        shrink:     width * config.space_shrink,
                        width:      width * config.space_width,
                        stretch:    width * config.space_stretch,
                        height:     0.
                    };
                },
        width:  parent.getBoundingClientRect().width,
        items:  items
    };
    
    let lines = run(context);
    let y = 0.;
    for (var line of lines) {
        let lineElement = document.createElement("line");
        lineElement.style.top = y + "px";
        
        let height = line[0];
        let words = line[1];
        for (var word of words) {
            let text = word[0];
            let x = word[1];
            
            let wordElement = document.createElement("span");
            wordElement.appendChild(document.createTextNode(text));
            wordElement.style.left = x + "px";
            lineElement.appendChild(wordElement);
        }
        y += height;
        
        parent.appendChild(lineElement);
    }
}

function max(a, b) {
    return a > b ? a : b;
}

function measure_at(measure, factor) {
    var d;
    if (factor < 0.) {
        d = measure.width - measure.shrink;
    } else {
        d = measure.stretch - measure.width;
    }
    
    return d * factor + measure.width;
}

function measure_add(a, b) {
    return {
        shrink:     a.shrink  + b.shrink,
        width:      a.width   + b.width,
        stretch:    a.stretch + b.stretch,
        height: max(a.height,   b.height)
    };
}

function measure_word(word) {
    
}
function measure_space(scale) {
    
}

function run(self) {
    let limit = self.items.length;
    let nodes = [{
        prev:   0,
        path:   0,
        factor: 0.,
        score:  0.
    }];
    let last = 0;
        
    for (var start = 0; start < limit; start++) {  
        let b = nodes[start];
        if (b != undefined) {
            last = complete_line(self, nodes, {
                measure: {
                    shrink:     0.,
                    width:      0.,
                    stretch:    0.,
                    height:     0.
                },
                path:       0,
                score:      b.score,
                begin:      start,
                pos:        start,
                branches:   0
            });
        }
    }
    
    let steps = [];
        
    while (last > 0) {
        let b = nodes[last];
        steps.push([b, last-1]);
        last = b.prev;
    }
        
    let lines = [];
    for (var step of steps.reverse()) {
        let b = step[0];
        let end = step[1];
        
        let measure = {
            shrink:     0.,
            width:      0.,
            stretch:    0.,
            height:     0.
        };
        let words = [];
        let pos = b.prev;
        let branches = 0;
        while (pos < end) {
            let node = self.items[pos];
            switch (node[0]) {
                case 0: // Word
                    let w = node[1];
                    let x = measure_at(measure, b.factor);
                    measure = measure_add(measure, self.word(w));
                    words.push([w, x]);
                    break;
                
                case 2: // Space
                    let s = node[2];
                    measure = measure_add(measure, self.space(s));
                    break;
                
                case 2: // BranchEntry
                    let len = node[1];
                    if (b.path & (1<<branches) == 0) {
                        pos += len;
                    }
                    branches += 1;
                    break;
                
                case 3: // BranchExit
                    let skip = node[1];
                    pos += skip;
                    break;
            }
            pos += 1;
        }
        
        lines.push([measure.height, words]);
    }
    return lines;
}
    
function complete_line(self, nodes, c) {
    let last = c.begin;
    
    while (c.pos < self.items.length) {
        let n = c.pos;
        let item = self.items[n];
        switch (item[0]) {
            case 0: // Word
                let w = item[1];
                c.measure = measure_add(c.measure, self.word(w));
                break;
            
            case 2: // Space
                let breaking = item[1];
                let s = item[2];
                if (breaking) {
                    if (maybe_update(self, c, nodes, n+1)) {
                        last = n+1;
                    }
                }
                
                // add width now.
                c.measure = measure_add(c.measure, self.space(s));
                break;
            
            case 1: // Linebreak
                let fill = item[1];
                if (fill) {
                    if (self.width > c.measure.stretch) {
                        c.measure.stretch = self.width;
                        if (self.width > c.measure.width) {
                            c.measure.width = self.width;
                        }
                    }
                }
            
                if (maybe_update(self, c, nodes, n+1)) {
                    last = n+1;
                }
                return last;
            
            case 2: // BranchEntry
                let b_last = complete_line(self, nodes, {
                    pos:        n + 1,
                    path:       c.path | (1 << c.branches),
                    branches:   c.branches + 1,
                    score:      b.score,
                    begin:      start,
                    measure:    measure
                });
                last = max(last, b_last);
                
                c.pos += len;
                c.branches += 1;
                break;
            
            case 3: // BranchExit
                let skip = item[1];
                c.pos += skip;
                break;
        }
        
        if (c.measure.shrink > self.width) {
            return last; // too full
        }
        
        c.pos += 1;
    }
    
    return last;
}


function maybe_update(self, c, nodes, index) {
    let m = c.measure;
    let width = self.width;
    
    if (width < m.shrink) {
        return false;
    }
    
    var factor;
    if (width == m.width) {
        factor = 1.0;
    } else {
        let delta = width - m.width; // d > 0 => stretch, d < 0 => shrink
        let diff;
        if (delta >= 0.) {
            diff = m.stretch - m.width;
        } else {
            diff = m.width - m.shrink;
        }
        factor = delta / diff;
    };
    let break_score = c.score - factor * factor;
    let other = nodes[index];
    if (other == undefined) {
        nodes[index] = {
            prev:   c.begin,
            path:   c.path,
            factor: factor,
            score:  break_score
        };
    } else if (break_score > other.score) {
        nodes[index] = {
            prev:   c.begin,
            path:   c.path,
            factor: factor,
            score:  break_score
        };
    }
    return true;
}
